Update integration notebook (#662)
AlexsLemonade/training-modules: scRNA-seq-advanced/03-dataset_integration.Rmd

---

In this notebook, we'll perform integration on scRNA-seq datasets from the [Single-cell Pediatric Cancer Atlas (`ScPCA`)](https://scpca.alexslemonade.org/), a database of uniformly-processed pediatric scRNA-seq data built and maintained by the Data Lab.

> The `ScPCA` database currently hosts single-cell pediatric cancer transcriptomic data generated by ALSF-funded labs, with a leading goal of making this data easily accessible to investigators (like you!).

Reviewer: I don't love "leading goal"... "primary goal", "main goal"? None of those sound quite right. I have "the" goal below, because just "a goal" sounded bad. 
```suggestion
The `ScPCA` database currently hosts single-cell pediatric cancer transcriptomic data generated by ALSF-funded labs, with the goal of making this data easily accessible to investigators (like you!).
```

The expression data in `ScPCA` were mapping and quantified with [`alevin-fry`](https://doi.org/10.1038/s41592-022-01408-3), followed by processing with Bioconductor tools using the same general procedures that we have covered in this workshop.

> The raw counts matrix was filtered with `emptyDropsCellRanger()` and `miQC` and log-normalized with `scuttle`, with dimension reduction performed by `scater`.

Reviewer: and -> Comma
```suggestion
The raw counts matrix was filtered with `emptyDropsCellRanger()` and `miQC`, log-normalized with `scuttle`, with dimension reduction performed by `scater`.
```

---

In this notebook, we'll perform integration on scRNA-seq datasets from the [Single-cell Pediatric Cancer Atlas (`ScPCA`)](https://scpca.alexslemonade.org/), a database of uniformly-processed pediatric scRNA-seq data built and maintained by the Data Lab.

> This database currently hosts (as of March 2023) eleven projects from eight different ALSF-funded labs who have generated single-cell pediatric cancer transcriptomic data with the ultimate goal of making this data easily accessible to investigators (like you!).
> `ScPCA` raw data is processed with [nextflow pipeline](https://github.com/AlexsLemonade/scpca-nf) that uses [`alevin-fry`](https://doi.org/10.1038/s41592-022-01408-3) for fast and efficient mapping, followed by filtering and normalization with Bioconductor tools. 
> You can read more about how data in the `ScPCA` is processed in [the associated documentation](https://scpca.readthedocs.io/en/latest/).

Reviewer: I was really looking for both more and less here... I was mostly hoping to just get across that we used the most of same tools that they have seen in the first notebook. I added some of that detail below, but check that it is correct!

I also don't think the stats are needed here, especially if we have to keep updating them, so I went the opposite direction and cut the numbers. I also realized the "ultimate goal" part was ambiguous (who's goal?). So I tried to rephrase a bit.

```suggestion
The `ScPCA` database currently hosts single-cell pediatric cancer transcriptomic data generated by ALSF-funded labs, with the ultimate goal of making this data easily accessible to investigators (like you!).
The expression data in `ScPCA` were mapping and quantified with [`alevin-fry`](https://doi.org/10.1038/s41592-022-01408-3), followed by processing with Bioconductor tools, following the same general procedures that we have covered in this workshop.
The raw counts matrix was filtered with `emptyDropsCellRanger()` and `miQC`, log-normalized with `scuttle`, with dimension reduction by `scater`.
The processed data is stored as `.rds` files containing `SingleCellExperiment` objects.
You can read more about how data in the `ScPCA` is processed in [the associated documentation](https://scpca.readthedocs.io/en/latest/).
```

In this notebook, we'll perform integration on scRNA-seq datasets from the [Single-cell Pediatric Cancer Atlas (`ScPCA`)](https://scpca.alexslemonade.org/), a database of uniformly-processed pediatric scRNA-seq data built and maintained by the Data Lab.
This database currently hosts projects from ten different ALSF-funded labs who have generated single-cell pediatric cancer transcriptomic data with the ultimate goal of making this data easily accessible to investigators (like you!).

> **MORE TEXT HERE ON SCPCA PROCESSING**

Reviewer: Just to be clear, this can be a sentence that says the data was mapped with alevin fry, then filtered and normalized... If you want to link to the docs, you can do that, but I don't think we need much detail at all!

Integration corrects for batch effects that arise from different library preparations, genetic backgrounds, and other sample-specific factors, so that datasets can be jointly analyzed at the cell level. 
`fastMNN` corrects for batch effects using a faster variant of the mutual-nearest neighbors algorithm, the technical details of which you can learn more about from this [vignette by Lun (2019)](https://marionilab.github.io/FurtherMNN2018/theory/description.html).
`harmony`, on the other hand, corrects for batch effects using an iterative clustering approach, and unlike `fastMNN`, it is also able to consider additional covariates beyond just the batch groupings.

Regardless of which integration tool is used, the `SingleCellExperiment` (SCE) objects first need to be reformatted and merged into a single (uncorrected!) SCE object that contains all cells from all samples.
This merged SCE can then be used for integration to obtain a formally batch-corrected SCE object.

> ![Single-cell roadmap: Integrate](diagrams/roadmap_multi_integrate.png)

Reviewer: I had a second, so here is a version of the diagram I am suggesting: https://docs.google.com/presentation/d/1j_qH4KD2WxAbwX3n9vCkoiG87E1wIhnYzThjk8VgZ34/edit#slide=id.g21270a3c171_0_197

> ![Single-cell roadmap: Integration Overview](diagrams/roadmap_multi_merge-integrate.png)

Reviewer: With this change, the other part of my comment seems to have been hidden on this page, so I am adding it again here:

> In conjunction with this diagram (either version), I think it will be worth adding a note about what processing has already been done on these samples. i.e., that they have already been filtered and normalized.

I also think we could now move this diagram up higher in this notebook...after the first paragraph probably makes sense, and that would be a good place to also add a bit more text about ScPCA processing that has been done.

A different and more modular coding approach to reading in these files (and more!) is to leverage the [`purrr`](https://purrr.tidyverse.org/) `tidyverse` package, which provides a convenient set of functions for operating on lists.
You can read more about these functions and their power and utility in R in [Chapter 21 of _R for Data Science_](https://r4ds.had.co.nz/iteration.html#for-loops-vs.-functionals).

Of particular interest is the [`purrr::map()`](https://purrr.tidyverse.org/reference/map.html) family of functions, which can be used to run a given function on each element of a list (or vector) in one call.
The general syntax for `purrr::map()` and friends is:

```
# Syntax for using the map function:
purrr::map(input list or vector, 
           function to apply to each item in the input, 
           any additional arguments to the function can go here, 

>            and also here if there are even more arguments, and so on)

Reviewer: I didn't notice this change, but I would probably revert the removal of the angle brackets. Having the bare text seems confusing to me.

```

The output from running `purr::map()` is always a list (but note that there are other `purrr::map()` relatives which return other object types, as you can read about in [the `purrr::map()` documentation](https://purrr.tidyverse.org/reference/index.html)).
If this concept sounds a little familiar to you, that's because it probably is!

> In fact, this is conceptually the same as base R's `lapply` function, and the `purrr::map()` family of functions provide more robust and reliable alternatives to the base R `apply` functions.

Reviewer: I'm trying to hedge this a bit... I don't think `purrr` provides a good alternative to `apply()` itself!

```suggestion
Base R's `lapply()` function can provide similar utility, and the `purrr::map()` family of functions can in part be thought of as an alternative to some of the base R `apply` functions, with more consistent behavior.
```

> Base R's `lapply()` function can provide similar utility, and the `purrr::map()` family of functions can in part be thought of as an alternative to some of the base R `apply` functions, with more consistent behavior.

Reviewer: Making "in part" parenthetical
```suggestion
Base R's `lapply()` function can provide similar utility, and the `purrr::map()` family of functions can (in part) be thought of as an alternative to some of the base R `apply` functions, with more consistent behavior.
```

```

We can now read these files in and create a list of four SCE objects.
Since `readr::read_rds()` can only operate on one input at a time, we'll need to use `purrr::map()` to run it on all input file paths in one command.
Although `sce_paths` is a vector (not a list), it will still work as input to `purrr:map()`.

> But, the output from this code will be a list, since that's what `purrr::map()` always returns.
> (We could convert `sce_paths` to a list first with `as.list(sce_paths)`, but for this example, the code will work just the same!)

Reviewer: I don't think adding a suggestion of something not to do is particularly helpful here. I also don't like starting with a "but" if we don't really need to!
```suggestion
The output from this code will still be a list, since that's what `purrr::map()` always returns.
```

## Prepare the SCE list for integration

> ![Single-cell roadmap: Merge](diagrams/roadmap_multi_merge.png)

Reviewer: Whether we need the merge diagram here again is up to you. I don't think it is terrible to have it twice, but we probably don't need to, as we will have a second integration diagram later. Or if you take the suggestion to change the initial diagram, leaving this here makes sense!
